File: waveforms.js
// this is waveforms.js
const tValues = new Array(800).fill(0).map((_, index) => (index / 800) * 5 * Math.PI);
const mu = 2 * Math.PI;
const sigma = 0.2;
const amplitude = 3;

function gaussian(x, mu, sigma) {
    return Math.exp(-((x - mu) ** 2) / (2 * sigma ** 2));
}

function refinedRWave(t, mu, sigma, amplitude) {
    return amplitude * gaussian(t, mu, sigma);
}

function generateFullWaveform(t, qrsVariationFunction = null) {
    const pWave = (0 <= t && t < Math.PI) ? Math.sin(t) ** 2 : 0;
    const qrs = (qrsVariationFunction === null) ? refinedRWave(t, mu, sigma, amplitude) : qrsVariationFunction(t);
    const sWave = (2.5 * Math.PI <= t && t < 3 * Math.PI) ? -0.5 * Math.sin(t - 2.5 * Math.PI) ** 3 : 0;
    const tWave = (3 * Math.PI <= t && t < 4 * Math.PI) ? 0.5 * Math.sin(t - 3 * Math.PI) ** 2 : 0;
    return 300 - (pWave + qrs + sWave + tWave) * -100;
}

// Variations
function qrsVariation1(t) {
    const pWave = (0 <= t && t < Math.PI) ? Math.sin(t) ** 2 : 0;
    const qrs = refinedRWave(t, mu, 4*sigma, 0.5*amplitude);
    const sWave = (2.5 * Math.PI <= t && t < 3 * Math.PI) ? -0.5 * Math.sin(t - 2.5 * Math.PI) ** 3 : 0;
    const tWave = (3 * Math.PI <= t && t < 4 * Math.PI) ? 0.5 * Math.sin(t - 3 * Math.PI) ** 2 : 0;
    return 300 - (pWave + qrs + sWave + tWave) * -100;
}

function qrsVariation2(t) {
    const pWave = (0 <= t && t < Math.PI) ? Math.sin(t) ** 2 : 0;
    const qrs = refinedRWave(t, 1.3 * mu, sigma, 0.8*amplitude); // Different QRS complex
    const sWave = (2.5 * Math.PI <= t && t < 3 * Math.PI) ? -0.5 * Math.sin(t - 2.5 * Math.PI) ** 3 : 0;
    const tWave = (3 * Math.PI <= t && t < 4 * Math.PI) ? 0.5 * Math.sin(t - 3 * Math.PI) ** 2 : 0;
    return 300 - (pWave + qrs + sWave + tWave) * -100;
}

function noPWave(t) {
    const pWave = 0;
    const qrs = refinedRWave(t, mu, sigma, amplitude);
    const sWave = (2.5 * Math.PI <= t && t < 3 * Math.PI) ? -0.5 * Math.sin(t - 2.5 * Math.PI) ** 3 : 0;
    const tWave = (3 * Math.PI <= t && t < 4 * Math.PI) ? 0.5 * Math.sin(t - 3 * Math.PI) ** 2 : 0;
    return 300 - (pWave + qrs + sWave + tWave) * -100;
}

function noSWave(t) {
    const pWave = (0 <= t && t < Math.PI) ? Math.sin(t) ** 2 : 0;
    const qrs = refinedRWave(t, mu, sigma, amplitude);
    const sWave = 0; // No S wave
    const tWave = (3 * Math.PI <= t && t < 4 * Math.PI) ? 0.5 * Math.sin(t - 3 * Math.PI) ** 2 : 0;
    return 300 - (pWave + qrs + sWave + tWave) * -100;
}

function noTWave(t) {
    const pWave = (0 <= t && t < Math.PI) ? Math.sin(t) ** 2 : 0;
    const qrs = refinedRWave(t, mu, sigma, amplitude);
    const sWave = (2.5 * Math.PI <= t && t < 3 * Math.PI) ? -0.5 * Math.sin(t - 2.5 * Math.PI) ** 3 : 0;
    const tWave = 0;
    return 300 - (pWave + qrs + sWave + tWave) * -100;
}

let waveforms = [
    tValues.map(t => generateFullWaveform(t)),
    tValues.map(t => generateFullWaveform(t, qrsVariation1)),
    tValues.map(t => generateFullWaveform(t, qrsVariation2)),
    tValues.map(t => noPWave(t)),
    tValues.map(t => noSWave(t)),
    tValues.map(t => noTWave(t))
];

function GenerateWaveform() {
    const abnormalWaveforms = [
        qrsVariation1, qrsVariation2,
        noPWave, noSWave, noTWave
    ];
    const waveformNames = [
        "QRS Variation 1", "QRS Variation 2",
        "No P wave", "No S wave", "No T wave"
    ];
    const isAbnormal = Math.random() < 0.6;
    if (!isAbnormal) {
        const waveformFunction = generateFullWaveform;
        const waveformType = 0; // Normal
        return [tValues.map(t => waveformFunction(t)), waveformType];
    } else {
        const index = Math.floor(Math.random() * abnormalWaveforms.length);
        const waveformFunction = abnormalWaveforms[index];
        const waveformType = 1; // Abnormal
        const newWaveform = tValues.map(t => waveformFunction(t));
        return [newWaveform, waveformType];
    }
}
// this is waveforms.js

File: waveforms.py
import numpy as np
import matplotlib.pyplot as plt

# ECG waveform generation functions without qrs_variation_3

t_values = np.linspace(0, 5 * np.pi, 800)
mu = 2 * np.pi
sigma = 0.2
amplitude = 3

def gaussian(x, mu, sigma):
    """Generates a Gaussian wave based on the provided parameters."""
    return np.exp(-((x - mu) ** 2) / (2 * sigma ** 2))

def refined_r_wave(t, mu, sigma, amplitude):
    """Generates the R-wave component of the ECG."""
    return amplitude * gaussian(t, mu, sigma)

def generate_full_waveform(t, qrs_variation_function=None):
    """Generates the full ECG waveform."""
    p_wave = np.sin(t)**2 if 0 <= t < np.pi else 0
    if qrs_variation_function is None:
        qrs = refined_r_wave(t, mu, sigma, amplitude)
    else:
        qrs = qrs_variation_function(t)
    s_wave = -0.5 * np.sin(t - 2.5 * np.pi)**3 if 2.5 * np.pi <= t < 3 * np.pi else 0
    t_wave = 0.5 * np.sin(t - 3 * np.pi)**2 if 3 * np.pi <= t < 4 * np.pi else 0
    return 300 - (p_wave + qrs + s_wave + t_wave) * -100

# Variations
def qrs_variation_1(t):
    return 1.8 * np.sin(t - 1.5 * np.pi)**3 if 1.5 * np.pi <= t < 2.5 * np.pi else 0

def qrs_variation_2(t):
    return 1.5 * np.sin(0.8 * (t - 1.5 * np.pi))**3 if 1.5 * np.pi <= t < 2.8 * np.pi else 0

def qrs_variation_1(t):
    pWave = np.sin(t) ** 2 if 0 <= t < np.pi else 0
    qrs = refined_r_wave(t, mu, 4 * sigma, 0.5 * amplitude)
    sWave = -0.5 * np.sin(t - 2.5 * np.pi) ** 3 if 2.5 * np.pi <= t < 3 * np.pi else 0
    tWave = 0.5 * np.sin(t - 3 * np.pi) ** 2 if 3 * np.pi <= t < 4 * np.pi else 0
    return 300 - (pWave + qrs + sWave + tWave) * -100

def qrs_variation_2(t):
    pWave = np.sin(t) ** 2 if 0 <= t < np.pi else 0
    qrs = refined_r_wave(t, 1.3 * mu, sigma, 0.8 * amplitude) # Different QRS complex
    sWave = -0.5 * np.sin(t - 2.5 * np.pi) ** 3 if 2.5 * np.pi <= t < 3 * np.pi else 0
    tWave = 0.5 * np.sin(t - 3 * np.pi) ** 2 if 3 * np.pi <= t < 4 * np.pi else 0
    return 300 - (pWave + qrs + sWave + tWave) * -100

def no_p_wave(t):
    p_wave = 0
    qrs = refined_r_wave(t, mu, sigma, amplitude)
    s_wave = -0.5 * np.sin(t - 2.5 * np.pi)**3 if 2.5 * np.pi <= t < 3 * np.pi else 0
    t_wave = 0.5 * np.sin(t - 3 * np.pi)**2 if 3 * np.pi <= t < 4 * np.pi else 0
    return 300 - (p_wave + qrs + s_wave + t_wave) * -100

def no_s_wave(t):
    p_wave = np.sin(t)**2 if 0 <= t < np.pi else 0
    qrs = refined_r_wave(t, mu, sigma, amplitude)
    s_wave = 0  # No S wave
    t_wave = 0.5 * np.sin(t - 3 * np.pi)**2 if 3 * np.pi <= t < 4 * np.pi else 0
    return 300 - (p_wave + qrs + s_wave + t_wave) * -100

def no_t_wave(t):
    p_wave = np.sin(t)**2 if 0 <= t < np.pi else 0
    qrs = refined_r_wave(t, mu, sigma, amplitude)
    s_wave = -0.5 * np.sin(t - 2.5 * np.pi)**3 if 2.5 * np.pi <= t < 3 * np.pi else 0
    t_wave = 0
    return 300 - (p_wave + qrs + s_wave + t_wave) * -100

waveforms = [
    [generate_full_waveform(t) for t in t_values],
    [generate_full_waveform(t, qrs_variation_1) for t in t_values],
    [generate_full_waveform(t, qrs_variation_2) for t in t_values],
    [no_p_wave(t) for t in t_values],
    [no_s_wave(t) for t in t_values],
    [no_t_wave(t) for t in t_values]
]

# this is waforms.py
# can you convert This into javascript?